### PR TITLE
MM-33402: Invalidated the LRU cache for channel members after moderat…

### PR DESCRIPTION
**Pull request by mkraft**
…ion changes.

#### Summary

* Fixed a bug that required a cache purge or server restart for channel moderation changes to be correctly applied. See the internal ticket for further details.
* Adds a test for the case described by the ticket.
* Refactors some code that iterates all of the channel member pages for use in two places.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-33402

#### Release Note

```release-note
Fixed a bug that required a cache purge or server restart for channel moderation changes to be correctly applied.
```

### CODE REVIEW

File: app/channel_test.go [hunk 1]
```diff
@@ -1901,6 +1904,34 @@ func TestPatchChannelModerationsForChannel(t *testing.T) {
...
+		}
+
+		mockSession := model.Session{UserId: user.Id}
+
+		_, err := th.App.PatchChannelModerationsForChannel(channel.DeepCopy(), addCreatePosts)
+		require.Nil(t, err)
+		require.True(t, th.App.SessionHasPermissionToChannel(mockSession, channel.Id, model.PERMISSION_CREATE_POST))
+
+		_, err = th.App.PatchChannelModerationsForChannel(channel.DeepCopy(), removeCreatePosts)
+		require.Nil(t, err)
```
Comment (streamer45) [inline app/channel_test.go:1932]: We now tend to use `NoError` instead of `Nil`.

File: app/channel.go [hunk 2]
```diff
@@ -3007,6 +3009,19 @@ func (a *App) ClearChannelMembersCache(channelID string) {
...
+
+func (a *App) ClearChannelMembersCache(channelID string) {
+	clearSessionCache := func(channelMember model.ChannelMember) error {
+		a.ClearSessionCacheForUser(channelMember.UserId)
+		message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_CHANNEL_MEMBER_UPDATED, "", "", channelMember.UserId, nil)
+		message.Add("channelMember", channelMember.ToJson())
+		a.Publish(message)
+		return nil
+	}
+	a.forEachChannelMember(channelID, clearSessionCache)
```
Comment (streamer45) [inline app/channel.go:3024]: Should we log the returned error (if any)?

**Review by streamer45 — APPROVED** [review]
Nice :+1: I really appreciated the functional vibe of the refactor.

File: app/channel_test.go (hunk 1 above)
Comment (mkraft) [reply app/channel_test.go:1932]: `(*App).PatchChannelModerationsForChannel` returns a `*model.AppError`, and I think it's like this:

* `error` interfaces should `require.Error` or `require.NoError`, and
* `model.AppError` interfaces should `require.Nil` or `require.NotNil`

Unless I'm wrong?

File: app/channel.go (hunk 2 above)
Comment (mkraft) [reply app/channel.go:3024]: Nothing in the `clearSessionCache` function returns an error, so it'll never return one in this case.

Comment (streamer45) [reply app/channel.go:3024]: I was thinking about `GetMembers()` failing. Before the refactor we were logging that.

File: app/channel_test.go (hunk 1 above)
Comment (streamer45) [reply app/channel_test.go:1932]: Oh interesting. I was convinced `Error/NoError` did actual checks on the interface but instead they merely check for nil equality. Looks good then :+1:

File: app/channel.go (hunk 2 above)
Comment (mkraft) [reply app/channel.go]: Good point. Added handling for it.

**Review by sbishel — APPROVED** [review]

**Review by furqanmlk — APPROVED** [review]
image
image

Found different issue while testing. unable to delete post reaction via API. I will log different issue for this,

**Pull request merged**